### PR TITLE
 Added BoundingBox.ContainsPrecise(BoundingFrustum) method

### LINKED ISSUE #6298: Error in BoundingBox.Contains(BoundingFrustum)
https://github.com/MonoGame/MonoGame/blob/6f4e51030120913bc1d5b21b34b07493d2eeebff/MonoGame.Framework/BoundingBox.cs#L71

The function ```BoundingBox.Contains(BoundingFrustum)``` doesn't work properly. If I have a BoundingFrustum that is outside a BoundingBox, the returned value will be ```ContainmentType.Intersects```.

If I instead call the equivalent (?) function on the ```BoundingFrustum.Contains(BoundingBox)```, I get the right result, namely ```ContainmentType.Disjoint```.
https://github.com/MonoGame/MonoGame/blob/6f4e51030120913bc1d5b21b34b07493d2eeebff/MonoGame.Framework/BoundingFrustum.cs#L179

#### What version of MonoGame does the bug occur on:
- MonoGame 3.7

#### What operating system are you using:
- Windows

#### What MonoGame platform are you using:

- DesktopGL, WindowsDX, NetCore

### PR DESCRIPTION
**Pull request by baudii**
Fixes #6298

### Description of Change

Implemented the ContainsPrecise method for BoundingBox, as suggested by Member (#8564)
- ContainsPrecise performs an exact containment check between a BoundingBox and
BoundingFrustum, avoiding false positives
- Implemented 6 unit tests covering various scenarios for the ContainsPrecise method
- Tests include cases for full containment (1), intersection (3), and disjoint (2) relationships

### CODE REVIEW

**Comment by baudii** [comment]
I've been having some discussions about performance of the original implementation, so I also include benchmarks comparing my method with original implementation.

If you don't want to open the .txt file, on average the difference is like this:

```
Original implementation
Total: 1018 ms for 10000000 iterations
Average: 0,00010187242 ms

ContainsPrecise
Total: 1387 ms for 10000000 iterations
Average: 0,00013878448 ms
```
[Benchmarks.txt](https://github.com/user-attachments/files/17693631/Benchmarks.txt)

Yes, actually calculating the result is slower, but not that much slower.

**Comment by squarebananas** [comment]
From your #8564 post:
> I read about this theorem when you first mentioned it. But I figured you don't really need it here.

I understand it would be great if we **_"don't really need it here"_**. Disproving a proven theorem would be quite an achievement as well. Consider this example though:

box-frustum

It is clear it isn't contained, but whether the result is intersect or disjoint still needs determining. Using your method (essentially converting the frustum to a axis aligned box) the result would be intersect right?

Does this match the result from using separating axis theorem, or will a hyperplane of separation be discovered?

<details>
<summary><h3><b>Answer</b></h3></summary>

Don't get me wrong, it's good you're trying to tackle this, it's just non axis aligned collision gets complex. Were you trying to get this working for a particular reason like a project you're working on by the way?
</details>

**Comment by baudii** [comment]
@squarebananas Damn, I really thought I had it... Yeah, I wasn't trying to dispove the theorem, it was supposed to be more of "let's not calculate the area of a circle by integrating", hehe.

**BUT!**, this implementation will actually provide proper false-positives in a non-toxic way, right? Unlike original implementation which basically does nothing and does it wrong.

Wish there were prewritten testcases to actually test my implementation considering such edge cases. Probably, there aren't any, because nobody actually uses this method. But still.

>Don't get me wrong, it's good you're trying to tackle this, it's just non axis aligned collision gets complex. Were you trying to get this working for a particular reason like a project you're working on by the way?

I took on this issue because it was marked as "Good First Issue". Now I went way too far to turn back 🤣

**Comment by squarebananas** [comment]
> **BUT!**, this implementation will actually provide proper false-positives in a non-toxic way, right? Unlike original implementation which basically does nothing and does it wrong.

Yeah. although it may provide more false positives than the Frustum to Box method where plane checks are used.

The thing to note is in many use cases the frustum is large and the box is small. Looking up at the first example in the first image, if you drew a rectangle around the frustum then almost half of the area is outside of the frustum. As a result I would expect far more false-positives compared to the planes check method (the one the Frustum with Box method already uses).

For cases where the box is much larger than the frustum, I would expect that the opposite is true and your code here would work better. An experiment for a later date maybe.

> Wish there were prewritten testcases to actually test my implementation considering such edge cases. Probably, there aren't any, because nobody actually uses this method. But still.

Frustum culling gets used a lot in 3d graphics including MonoGame. The thing is false negatives are very visible (objects disappearing when in view) whereas false positive issues are not immediately apparent (objects being submitted for drawing even though they're not in view). Most developers are likely not aware of the issue and the potential impact it could have, especially if for example a high polygon terrain chunk is filling the GPU bandwidth even though it's fully behind the camera.

> I took on this issue because it was marked as "Good First Issue". Now I went way too far to turn back 🤣

Yeah, it shouldn't have had that label in my opinion. Frustums seem the most complex out of the MonoGame bounding types. There is still time to run from this.

Also warning unit tests for incorrect collision results could still be added. That way it makes it clearer this issue will need addressing. Then when it is fixed, those unit tests can be changed from warning to fail, ensuring such collisions methods won't be broken at a later date.

**Comment by baudii** [comment]
@squarebananas Been busy for the last few days. Can finally get back to this problem. 

Can you plz tell me if this is precise? I believe its mathematically correct. 

```csharp
public ContainmentType ContainsPrecise(BoundingFrustum frustum)
{
    Vector3[] boxNormals =
    [
        Vector3.Up,
        Vector3.Right,
        Vector3.Forward
    ];

    Vector3[] frustumNormals =
    [
        frustum.Left.Normal,
        frustum.Right.Normal,
        frustum.Top.Normal,
        frustum.Bottom.Normal,
        frustum.Far.Normal
    ];

    // allNormals = box normals + frustum normals + cross products of box normals and frustum normals
    Vector3[] allNormals = new Vector3[23]; // 3 + 5 + 3 * 5

    allNormals[0] = boxNormals[0];
    allNormals[1] = boxNormals[1];
    allNormals[2] = boxNormals[2];

    for (int i = 0; i < frustumNormals.Length; i++)
    {
        allNormals[3 + i] = frustumNormals[i];
        for (int j = 0; j < boxNormals.Length; j++)
        {
            allNormals[8 + i * boxNormals.Length + j] = Vector3.Cross(frustumNormals[i], boxNormals[j]);
        }
    }

    var boxCorners = GetCorners();
    var frustumCorners = frustum.GetCorners();

    bool intersects = false;

    for (int i = 0; i < allNormals.Length; i++)
    {
        float boxMin = float.MaxValue, boxMax = float.MinValue;
        float frustumMin = float.MaxValue, frustumMax = float.MinValue;

        foreach (var point in boxCorners)
        {
            var dot = Vector3.Dot(point, allNormals[i]);
            if (boxMin > dot) boxMin = dot;
            if (boxMax < dot) boxMax = dot;
        }

        foreach (var point in frustumCorners)
        {
            var dot = Vector3.Dot(point, allNormals[i]);
            if (frustumMin > dot) frustumMin = dot;
            if (frustumMax < dot) frustumMax = dot;
        }

        if (boxMax < frustumMin || boxMin > frustumMax)
            return ContainmentType.Disjoint;
        else if (boxMax < frustumMax || boxMin > frustumMin)
            intersects = true;
    }

    if (intersects)
        return ContainmentType.Intersects;

    return ContainmentType.Contains;
}
```

**Comment by baudii** [comment]
>Just one minor thing I noticed regards the summary comment "but might affect performance"

Sure, I'll add that. If you have any optimization advices I'd love to hear them. The only thing I see is I could remove the `boxNormals` array and use `allAxes` instead or switch `foreach` with `for`, but it's so minor that I feel like leaving it as it is just for the sake of readabillity

>Also BoundingBox.ContainsPrecise(BoundingFrustum) could be used to add a BoundingFrustum.ContainsPrecise(BoundingBox) method as well (following the pattern of all other collision methods). I think the only difference would be swapping the < and > around on the (boxMax < frustumMax || boxMin > frustumMin) line for the contains condition.

If the math is correct, since SAT is true for any convex shape, you could actually with small adjustment make this method static and abstract it for every convex shape by putting it in some sort of `ShapeHelper.cs` or something class if there is one. Like this:
```csharp
public static ContainmentType ContainsPreciseSAT(Vector3[] shape1Normals, Vector3[] shape2Normals,
                                                 Vector3[] shape1Corners, Vector3[] shape2Corners)
{
    // Check if shape1 contains, intersects or disjoint from shape2
    int l1 = shape1Normals.Length;
    int l2 = shape2Normals.Length;

    Vector3[] allAxes = new Vector3[l1 + l2 + (l1 * l2)];

    for (int i = 0; i < l1; i++)
    {
        allAxes[i] = shape1Normals[i];
    }

    for (int i = 0; i < l2; i++)
    {
        allAxes[l1 + i] = shape2Normals[i];
        for (int j = 0; j < l1; j++)
        {
            allAxes[l1 + l2 + i * l1 + j] = Vector3.Cross(shape2Normals[i], shape1Normals[j]);
        }
    }

    bool intersects = false;

    for (int i = 0; i < allAxes.Length; i++)
    {
        // Project both shapes on the axis
        float shape1Min = float.MaxValue, shape1Max = float.MinValue;
        float shape2Min = float.MaxValue, shape2Max = float.MinValue;

        foreach (var point in shape1Corners)
        {
            var dot = Vector3.Dot(point, allAxes[i]);
            if (shape1Min > dot) shape1Min = dot;
            if (shape1Max < dot) shape1Max = dot;
        }

        foreach (var point in shape2Corners)
        {
            var dot = Vector3.Dot(point, allAxes[i]);
            if (shape2Min > dot) shape2Min = dot;
            if (shape2Max < dot) shape2Max = dot;
        }

        if (shape1Max < shape2Min || shape1Min > shape2Max)
            return ContainmentType.Disjoint;
        else if (shape1Max < shape2Max || shape1Min > shape2Min)
            intersects = true;
    }

    if (intersects)
        return ContainmentType.Intersects;

    return ContainmentType.Contains;
}
```

But that sounds like a problem for another issue

>If it's alright I'll try to put a sample MonoGame project together for visually checking each bounding type. I have a 3d level editor I wrote years ago which I can adapt for this purpose. It's just I might not be able to sort that until next month if that's ok.

Thanks. Sounds good.

**Comment by baudii** [comment]
@squarebananas Yo. What’s up? Have you tested it?

**Comment by squarebananas** [comment]
> @squarebananas Yo. What’s up? Have you tested it?

I was hoping to get my 3d level collision editor (see example image) adapted to show each bounding type earlier in the month. My idea was to use it not only to test this PR but all combinations of existing bounding types & methods to find what works and doesn't. I did start adapting it and was actually looking at it a bit today. I'm still hoping to sort it this month but I'll test out your SAT method sooner in any case.

image

**Comment by squarebananas** [comment]
I found some interesting code yesterday within Microsoft's DirectXMath library which I want to look into. As I mentioned before there are lots of collision issues with MonoGame bounding volumes, so it would be useful to see how Microsoft are now handling collisions as well as try out their collision test samples. Until I do that I think I'm going to put adapting the 3d level editor on hold.

https://github.com/microsoft/DirectXMath/blob/main/Inc/DirectXCollision.inl
https://github.com/microsoft/Xbox-GDK-Samples/blob/main/Samples/System/Collision/media/image3.png

In the meanwhile I tested your two methods with my own game. It's testing all bounding boxes on screen against the camera frustum and also a mini frustum (100 times smaller), otherwise the intersect condition would be tested but the box fully contains frustum condition wouldn't. You can see the results in this video:

https://github.com/user-attachments/assets/680ef076-7481-4410-9347-f50e06cf78bc

The good news is no false positives or false negatives with your new methods. The original clearly has big issues with false positives (they are expected but not on this scale), but at least it doesn't have false negatives. (@SimonDarksideJ looks like you closed the issue (#6298) instead of the old PR (#7444) that had the wrong fix).

Also I managed to profile the CPU usage compared to `BoundingBox.Contains`. For my game I found that `BoundingBox.ContainsPrecise` is 1,350% more expensive and `ContainsPreciseSAT` is 1,600% more expensive. Like I mentioned this is to be expected, it's useful for a number of things like high polygon mesh culling and unit tests, but it shouldn't be used for mass object render culling. Now we know for certain the "might affect performance" summary should be updated to make this much clearer (see previous suggestions above).

**Comment by baudii** [comment]
> The good news is no false positives or false negatives with your new methods

Those are a good news indeed. So method is working correctly. Also, I fixed the summary as you suggested. 

_**`Offtopic`**_

>For my game I found that BoundingBox.ContainsPrecise is 1,350% more expensive and ContainsPreciseSAT is 1,600% more expensive

Hmmm... Interesting. It's interesting because `BoundingBox.ContainsPrecise` and `ContainsPreciseSAT` are practically the same method with the exact same logic. The only difference is that the SAT version is more abstract and can work with any convex shape by passing in the normals and edges. It's surprising that it increased performance costs by such a large margin. But it doesn't really matter. It was added just as an example to show that you can isolate the logic if its ever needed in the future. It's not a part of this PR.

>The original clearly has big issues with false positives (they are expected but not on this scale), but at least it doesn't have false negatives

It practically can't have false negatives, because there is no exit point that would return "Disjoint". But I'm actually curious about what it was intended to achieve by design. Should the implementation prioritize raw speed at the cost of accuracy, or aim for a balanced approach? Like, what could bring you better performance and somewhat acceptable accuracy with some false positives here and there but no false negatives? The only way to avoid false negatives without increasing performance is to remove them entirely, just like it's done in the original implementation. There literally can't be a more performant solution since it doesn't actually calculate anything. :)

But I guess other options could be to either calculate with planes as it is done in all other methods (by checking if all edges are on the same side of the plane) or use a method that I initially introduced here – essentially turning the frustum into a parallelepiped and checking for the collision of two bounding boxes. It would also help to benchmark these options to understand the trade-offs between performance and false positives.

But I guess I got a bit carried away and its also a topic for another PR that will focus on `Contains` method.

**Comment by squarebananas** [comment]
> It's surprising that it increased performance costs by such a large margin.

Compared to each other (instead of against `BoundingBox.Contains`) `ContainsPreciseSAT` takes 20% longer than `BoundingBox.ContainsPrecise`, so it's not a huge difference. It's actually what I was expecting, as with the `shapeNormals` parameters I literally used all shape normals. In reality the frustum's far and near normals are coincident vectors and don't need to be tested separately. Therefore it's 5 normals vs 6 normals being tested, the latter being 20% extra calculations. I did this because it's how I expect a typical user would probably use the function. The main point though was both SAT functions are precise but computationally expensive compared to the existing `BoundingBox.Contains`.

> It practically can't have false negatives, because there is no exit point that would return "Disjoint".

Right, I forgot how broken the existing `BoundingBox.Contains` method is. I should have noticed because in the video the `BoundingBox.Contains` false intersect count and the total disjoint count (at the top-right) are always the same. Currently `BoundingBox` always acts like the entire universe intersects with it when checked against any `BoundingFrustum`.

> Should the implementation prioritize raw speed at the cost of accuracy, or aim for a balanced approach?

Need to benchmark it like you say, but GPUs are generally more suited for parallel operations, so a few unneeded draw calls tend to be preferred over lots of extra CPU work. Basically it comes down to will making the CPU do more work be worth the reduction in GPU work.

Let's say making the `BoundingBox.Contains` slightly more accurate means the function will take 10% longer and there are 10,000 potential objects on a scene. If this results in 2,000 objects no longer being sent to the GPU every frame it might be worth it, whereas if it means a reduction of 20 objects then probably not.

I think benchmarking along with Microsoft's DirectXMath library would be useful as well. I only realised it contains collision functions yesterday, but I know the library is developed by former XNA staff as well as members of the current Xbox Advanced Technology Group. I expect it contains every possible optimized contains method there is.

**Comment by baudii** [comment]
Okay, can anyone review now?

**Comment by SimonDarksideJ** [comment]
@squarebananas / @baudii , do you both believe this is ready to go now?

**Review by SimonDarksideJ — CHANGES_REQUESTED** [review]
Just some code styling changes, otherwise looks good

File: MonoGame.Framework/BoundingBox.cs [hunk 1]
```diff
@@ -146,6 +146,89 @@ public ContainmentType Contains(BoundingFrustum frustum)
...
+
+                foreach (var point in frustumCorners)
+                {
+                    var dot = Vector3.Dot(point, allAxes[i]);
+                    if (frustumMin > dot) frustumMin = dot;
+                    if (frustumMax < dot) frustumMax = dot;
+                }
+
+                // If we find a gap, we are sure the shapes are disjoint
+                if (boxMax < frustumMin || boxMin > frustumMax)
```
[the remark this answers is not in the document]
Comment (squarebananas) [reply MonoGame.Framework/BoundingBox.cs:221]: > Braces are not used for single statement blocks <https://github.com/MonoGame/MonoGame/blob/develop/CODESTYLE.md#bracing>

?

**Comment by squarebananas** [comment]
> do you both believe this is ready to go now?

Already tested as shown in the video above, the post that says:
> (@SimonDarksideJ looks like you closed the issue (https://github.com/MonoGame/MonoGame/issues/6298) instead of the old PR (https://github.com/MonoGame/MonoGame/pull/7444) that had the wrong fix).

File: MonoGame.Framework/BoundingBox.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (baudii) [reply MonoGame.Framework/BoundingBox.cs:221]: So, should I add the braces? And if I do, should I add them everywhere or only these two places that were mentioned?

**Review by SimonDarksideJ — APPROVED** [review]
